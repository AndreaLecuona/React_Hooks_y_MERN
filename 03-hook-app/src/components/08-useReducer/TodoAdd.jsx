import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ( { handleAdd } ) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    })

    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1 ) {
            return
        };

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        //reemplazo este dispatch con handle add para que el dispatch se haga en el componente padre
        handleAdd( newTodo );
        reset();
    }

    return (
        <div className="form-group">
            <form onSubmit={ handleSubmit }>
                <input type="text"
                    name="description"
                    className="form-control"
                    placeholder="Tarea por realizar..."
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button type="submit"
                    className="btn btn-outline-primary personal-btn mt-3">
                    Agregar a la lista
                </button>
            </form>
        </div>
    )
}
