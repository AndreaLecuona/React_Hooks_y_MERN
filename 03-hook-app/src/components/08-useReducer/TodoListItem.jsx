import React from 'react'

export const TodoListItem = ( { todo, index, handleDelete, handleToggle } ) => {
    return (
        <li
            key={ todo.id }
            className="list-item"
        >
            <p 
                className={`${ todo.done && 'complete' }`}
                onClick={() => handleToggle( todo.id )}>

                    {index + 1}. { todo.desc }
            </p>

            <button 
                className="btn btn-danger personal-btn"
                onClick={() => handleDelete( todo.id )}>

                    Borrar
            </button>
        </li>
    )
}
