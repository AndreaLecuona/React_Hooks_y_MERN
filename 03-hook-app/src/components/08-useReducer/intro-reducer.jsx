const initialState = [{
    id: 1,
    todo: 'Comprar cafe',
    done: false
}];

//El use reducer centraliza todas las acciones posibles de la aplicacion, que van a afectar el state. No puede ser asíncrona, llamar al local storage, ni llamar a otras funciones desde su cuerpo o efectos secundarios (es pura)
//Recibe como argumentos un nuevo state y una accion(que a su vez puede o no tener un argumento) que va a modificar el initial state, y SIEMPRE devuelve un state. NUNCA NULL O UNDEFINED
const todoReducer = ( state = initialState, action ) => {

    if(action?.type === 'agregar nuevo item'){ //optional chaining: 'si action no es undefined o null, comparar con el string'
        return [ ...state, action.payload ];
    }

    return state;
};

let todos = todoReducer();


const newtodo = {
    id: 2,
    todo: 'Cocinar brownies',
    done: false
};

//Como .push() no se utiliza en react porque muta el objeto/modifica el state y rompe la app, tenemos que pasar el nuevo item de otra forma, junto con la acción
//en este objeto 'accion' que se enviará como argumento al reducer sse incluyen el nombre de la accion propiamente dicha y el objeto que involucra
//type y payload son nombres estandard, podrían ser los que quisiéramos
const agregarTodoAction = {
    type: 'agregar nuevo item',
    payload: newtodo
}


todos = todoReducer( todos, agregarTodoAction );


console.log(todos);