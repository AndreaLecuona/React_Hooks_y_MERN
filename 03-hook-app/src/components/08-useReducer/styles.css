h1{
    margin-left: 30px;
}
ul{
    list-style: none;
}
li{
    display: flex;
    justify-content: space-between;
    border-bottom: 0.5px solid black;
    padding: 2rem 0;
}
.form-group{
    width: 50%;
    margin-left: 30px;
}
