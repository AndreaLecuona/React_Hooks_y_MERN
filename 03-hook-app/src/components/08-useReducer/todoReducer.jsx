
export const todoReducer = (state = [], action) => {
    
    switch ( action.type ) {
        //debe conocer de antemano todas las posibles 'acciones' que le mandemos
        case 'add':
            return [...state, action.payload];

        case 'delete':
            return state.filter( todo => todo.id !== action.payload); //action-payload toma el id

        case 'toggle':
            return state.map( todo => 
                ( todo.id === action.payload ) ? { ...todo, done: !todo.done } : todo
            );

        //VERSION LARGA DE TOGGLE
        // case 'toggle':
        //     return state.map( todo => {
        //         if( todo.id === action.payload ){
        //             return {
        //                 ...todo,
        //                 done: !todo.done
        //             }
        //         } else {
        //             return todo;
        //         }
        //     })

        default:
            return state; //para asegurarnos que devuelva un state
    }

};