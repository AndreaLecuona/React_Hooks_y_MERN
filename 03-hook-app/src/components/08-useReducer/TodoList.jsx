import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ( {todos, handleDelete, handleToggle} ) => {
    return (
        <ul className="list-movies">
            {
                todos.map((todo, i) =>

                    <TodoListItem
                        key ={ todo.id }
                        todo ={ todo }
                        index ={ i }
                        handleDelete = { handleDelete }
                        handleToggle = { handleToggle }
                    />

                )
            }
        </ul>
    )
}
