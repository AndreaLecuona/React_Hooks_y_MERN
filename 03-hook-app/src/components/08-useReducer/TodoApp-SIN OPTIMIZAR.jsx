import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../hooks/useForm';

import './styles.css';

//const [state, dispatch] = useReducer(reducer, initialState, init)
//init: Es una funcion para inicializar el state en caso que Ã©ste sea procesado o haga varias acciones (retorna el initial state)
//dispatch: Ayuda a "disparar"/enviar las acciones hacia el reducer para volver a renderizar en caso de haberse producido un cambio en el state


// const initialState = [{
//     id: new Date().getTime(),
//     desc: "Barry Lyndon - Stanley Kubrick",
//     done: false
// }];

const init = () => {
    //LEER DE LOCAL STORAGE Y MANDARLO COMO INITIAL STATE
    return JSON.parse(localStorage.getItem( 'todos' )) || [];

    // return(
    //     [{
    //         id: new Date().getTime(), //simulacion de id
    //         desc: "Barry Lyndon - Stanley Kubrick",
    //         done: false
    //     }]
    // )
}


export const TodoApp = () => {
    
    //MANEJO DE INPUT CON CUSTOM HOOK    const [ formValues, handleInputChange] = useForm({...
    const [ { description }, handleInputChange, reset ] = useForm({
        description: ''
    })

    //Extraigo del resultado del reducer una lista de items
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);
    console.log( description );


    //GUARDAR EN LOCAL STORAGE
    //guardar en primer renderizado y cada vez que la lista cambie
    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify( todos ) );
    }, [todos]);

    const handleDelete = (todoId) => {
        console.log(todoId);

        const action = {
            type: 'delete',
            payload: todoId
        };

        dispatch( action );
    }

    const handleToggle = (todoId) => {
        dispatch( {
            type: 'toggle',
            payload: todoId
        })
    }


    //MANEJO DE SUBMIT
    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1 ) {
            return
        };

        //1. creo el nuevo item
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        //2. lo agrego a una accion y defino el tipo de accion
        const action = {
            type: 'add',
            payload: newTodo
        };

        //3. mando la accion al reducer con dispatch
        dispatch( action );
        reset();
    }

    return (
        <div>
            <h1>Pelis para ver en cuarentena ( {todos.length} )</h1>
            <hr/>

            <div className="row">

                <div className="col-7">
                    
                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, i) =>
                                {return <li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p className={ `${ todo.done && 'complete' }` } 
                                    onClick={ () => handleToggle(todo.id) 
                                    }>
                                        {i + 1}. {todo.desc}
                                    </p>
                                    <button className="btn btn-danger"
                                    onClick={() => handleDelete(todo.id)}
                                    >
                                        Borrar
                                    </button>
                                </li>}
                            )
                        }
                    </ul>

                </div>

                <div className="col-5">
                    
                    <h4>Me falta ver...</h4>
                    <hr/>

                    <form onSubmit={handleSubmit}>
                        <input type="text"
                            name="description"
                            className="form-control"
                            placeholder="peli - director"
                            autoComplete="off"
                            value={ description }
                            onChange={ handleInputChange }
                        />

                        <button type="submit"
                            className="btn btn-outline-primary mt-2">
                            Agregar
                        </button>
                    </form>

                </div>

            </div>
            
        </div>
    )
}
