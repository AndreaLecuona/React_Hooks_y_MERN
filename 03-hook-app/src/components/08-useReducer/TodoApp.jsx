import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
// import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

import './styles.css';

//const [state, dispatch] = useReducer(reducer, initialState, init)
//init: Es una funcion para inicializar el state en caso que Ã©ste sea procesado o haga varias acciones (retorna el initial state)
//dispatch: Ayuda a "disparar"/enviar las acciones hacia el reducer para volver a renderizar en caso de haberse producido un cambio en el state


// const initialState = [{
//     id: new Date().getTime(),
//     desc: "Barry Lyndon - Stanley Kubrick",
//     done: false
// }];

const init = () => {
    //LEER DE LOCAL STORAGE Y MANDARLO COMO INITIAL STATE
    return JSON.parse(localStorage.getItem( 'todos' )) || [];

    // return(
    //     [{
    //         id: new Date().getTime(), //simulacion de id
    //         desc: "Barry Lyndon - Stanley Kubrick",
    //         done: false
    //     }]
    // )
}


export const TodoApp = () => {

    //Extraigo del resultado del reducer una lista de items
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    //GUARDAR EN LOCAL STORAGE
    //guardar en primer renderizado y cada vez que la lista cambie
    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify( todos ) );
    }, [todos]);

    const handleDelete = (todoId) => {
        console.log(todoId);

        const action = {
            type: 'delete',
            payload: todoId
        };

        dispatch( action );
    }

    const handleToggle = (todoId) => {
        dispatch( {
            type: 'toggle',
            payload: todoId
        })
    }

    const handleAdd = ( newTodo ) => {

        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div className="wrapper my-5">

            <div className="row both-sides mx-5">
                <h1>TodoApp ({ todos.length })</h1>

                <div className="col-sm-12 col-lg-5 mt-5">
                    <TodoList
                        todos={ todos }
                        handleDelete={ handleDelete }
                        handleToggle={ handleToggle }
                    />

                </div>

                <div className="col-sm-12 col-lg-7 mt-5">            
                    <TodoAdd
                    handleAdd={ handleAdd }
                    />

                </div>

            </div>
            
        </div>
    )
}
