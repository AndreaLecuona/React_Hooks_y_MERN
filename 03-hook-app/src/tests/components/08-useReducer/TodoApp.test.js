import { act } from "@testing-library/react"
import { mount, shallow } from "enzyme"
import { TodoApp } from "../../../components/08-useReducer/TodoApp"
import { demoTodos } from "../../fixtures/simulacion-de-array-para-pruebas"

describe('pruebas en TodoApp', () => {
    
    const wrapper = shallow( <TodoApp/> );

    //para prueba del localStorage:
    //hacemos que el setItem sea una simulacion de jest
    Storage.prototype.setItem = jest.fn( () => {}) //se puede definir vacía

    test('debe renderizarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();

    })

    // !!!no sería tan necesaria porque ya lo evaluamos en todoAdd
    test('debe agregar un todo', () => {

        //capturo:
        //el mount sirve para cuando queremos probar nuestro componente en contexto, la app en general
        //tiene todos los mismo metodos que venios usando pero cambia el nivel en que es renderizada la aplicacion.
        //Shallow en más básico, toma sólo el componente aislado. Mount considera sus componentes hijos y el contexto en que se inserta y actúa.
        const wrapper = mount( <TodoApp/> );

        //simulacion de agregar un todo:
        //busco a su componente hijo TodoAdd y en particular su prop handleAdd para ejecutar esa función con el primer item de mi lista simulada (luego lo repito con un segundo item)
        //hay que envolverlo en un act porque ejecutar handleAdd provoca la modificación de un state 
        act( () => {
          
            wrapper.find( 'TodoAdd' ).prop( 'handleAdd' )(demoTodos[0]);
            wrapper.find( 'TodoAdd' ).prop( 'handleAdd' )(demoTodos[1]);

        })

        //aserciones:
        //espero que el h3 muestre el lenght de acuerdo a la cantidad de items agregados
        expect( wrapper.find( 'h3' ).text().trim() ).toBe( 'Pelis por ver: 2' );

        //espero que los items se graben en localStorage
        //no evaluamos que se haya grabado efectivamente, sólo que se haya llamado al método para grabar con nuestra info
        expect( localStorage.setItem ).toHaveBeenCalledTimes( 2 );

    })

    // !!!no sería tan necesaria porque ya lo evaluamos en todoAdd
    test('debe eliminar un todo', () => {
        
        //usando como referencia el wrapper con shallow, repetimos la simulación de agregar un todo
        wrapper.find( 'TodoAdd' ).prop( 'handleAdd' )(demoTodos[0]);
        expect( wrapper.find( 'h3' ).text().trim() ).toBe( 'Pelis por ver: 1' );

        //ejecutamos la accion de borrar
        wrapper.find( 'TodoList' ).prop( 'handleDelete' )(demoTodos[0].id);
        
        //esperamos que el lenght haya cambiado
        expect( wrapper.find( 'h3' ).text().trim() ).toBe( 'Pelis por ver: 0' );

    })
    

})

//NOTAS DE PROFESOR ASISTENTE
// Para las tareas de renderizado, eventos de usuario o data fetch que interactúan con la interfaz del usuario 
// es necesario el uso de act que se asegura que todas estas interacciones se hayan procesado y aplicado al DOM.
// Esto ayuda a que sus pruebas se acerquen más a lo que experimentarían los usuarios reales al usar su aplicación.

// Act() lo puedes usar con shallow o con mount() que permiten la renderización de componentes.

// al usar mount, react requiere que las actualizaciones en la interfaz sean envueltas en act() para manejar con mayor control los estados de actualización.
