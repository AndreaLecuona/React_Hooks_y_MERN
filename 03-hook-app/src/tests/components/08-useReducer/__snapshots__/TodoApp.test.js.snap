// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en TodoApp debe renderizarse correctamente 1`] = `
<div
  className="wrapper"
>
  <div
    className="row both-sides"
  >
    <div
      className="col-sm-12 col-lg-7"
    >
      <h1>
        Home cinema
      </h1>
      <h2>
        Quarantine edition
      </h2>
      <h3>
        Pelis por ver: 
        0
      </h3>
      <TodoAdd
        handleAdd={[Function]}
      />
    </div>
    <div
      className="col-sm-12 col-lg-5"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
  </div>
</div>
`;
